turbopack/crates/turbo-tasks-malloc/src/counter.rs

```diff
@@ -1,124 +1,198 @@
-/// size.
-const TARGET_BUFFER: usize = 100 * KB;
-/// When the thread-local buffer would exceed this size, we will update the
-/// global counter.
-const MAX_BUFFER: usize = 200 * KB;
+lazy_static::lazy_static! {
+    static ref ACTIVE: Mutex<GlobalData> = Mutex::new(Default::default());
+}
 
 #[derive(Default)]
+struct GlobalData {
+    /// Data accumulated from retired threads
+    allocations: usize,
+    deallocations: usize,
+    allocation_count: usize,
+    deallocation_count: usize,
+    active: FxHashSet<RegisteredCounters>,
+}
+
+struct AtomicCounters {
+    allocations: AtomicUsize,
+    deallocations: AtomicUsize,
+    allocation_count: AtomicUsize,
+    deallocation_count: AtomicUsize,
+}
+
+impl AtomicCounters {
+    const fn new() -> Self {
+        Self {
+            allocations: AtomicUsize::new(0),
+            deallocations: AtomicUsize::new(0),
+            allocation_count: AtomicUsize::new(0),
+            deallocation_count: AtomicUsize::new(0),
+        }
+    }
+
+    fn from_allocation_counters(counter: AllocationCounters) -> Self {
+        Self {
+            allocations: AtomicUsize::new(counter.allocations),
+            deallocations: AtomicUsize::new(counter.deallocations),
+            allocation_count: AtomicUsize::new(counter.allocation_count),
+            deallocation_count: AtomicUsize::new(counter.deallocation_count),
+        }
+    }
+    fn to_allocation_counters(&self) -> AllocationCounters {
+        AllocationCounters {
+            allocations: self.allocations.load(Ordering::Relaxed),
+            deallocations: self.deallocations.load(Ordering::Relaxed),
+            allocation_count: self.allocation_count.load(Ordering::Relaxed),
+            deallocation_count: self.deallocation_count.load(Ordering::Relaxed),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Eq, PartialEq, Hash, Copy, Clone)]
+struct RegisteredCounters(*const AtomicCounters);
+unsafe impl Send for RegisteredCounters {}
+unsafe impl Sync for RegisteredCounters {}
+
 struct ThreadLocalCounter {
-    /// Thread-local buffer of allocated bytes that have been added to the
-    /// global counter desprite not being allocated yet. It is unsigned so that
-    /// means the global counter is always equal or greater than the real
-    /// value.
-    buffer: usize,
-    allocation_counters: AllocationCounters,
+    counters: AtomicCounters,
+    registered: bool,
 }
 
 impl ThreadLocalCounter {
     const fn new() -> Self {
         Self {
-            buffer: 0,
-            allocation_counters: AllocationCounters::new(),
+            registered: false,
+            counters: AtomicCounters::new(),
         }
     }
+    #[inline]
+    #[cold]
+    fn register(&mut self) {
+        let ptr = RegisteredCounters(&self.counters as *const AtomicCounters);
+        let inserted = ACTIVE.lock().unwrap().active.insert(ptr);
+        debug_assert!(inserted);
```

Comment:
Maybe add an debug assert that that ACTIVE.len() < 100. That way we find cases where threads are not correctly unloaded.